src/crystal/system/win32/file_descriptor.cr

```diff
@@ -274,6 +273,47 @@ module Crystal::System::FileDescriptor
     end
     ret
   end
+
+  private def console_mode?(handle)
+    LibC.GetConsoleMode(handle, out _) != 0
+  end
+
+  private def read_console(handle : LibC::HANDLE, buffer : UInt16*, count : Int) : Int32
```

Comment:
Minor detail to consider: `buffer` and `count` could be collapsed into a single parameter of type `Slice(UInt16)`. `Slice` is the idiomatic data type for representing a pointer plus size.

The calculation for `units_buffer` would implicitly include the size with this expression: `(slice + third).unsafe_slice_of(UInt16)[0..-2]`